nx_cugraph/algorithms/link_prediction.py

```diff
@@ -0,0 +1,99 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import cupy as cp
+import networkx as nx
+import pylibcugraph as plc
+
+from nx_cugraph.convert import _to_undirected_graph
+from nx_cugraph.utils import networkx_algorithm, not_implemented_for
+
+__all__ = [
+    "jaccard_coefficient",
+]
+
+
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+@networkx_algorithm(version_added="25.02", _plc="jaccard_coefficients")
+def jaccard_coefficient(G, ebunch=None):
+    if ebunch is None:
+        # FIXME: is there a more efficient way to do this (on GPU or
+        # otherwise)?
+        ebunch = list(nx.non_edges(G))
+        if not ebunch:
+            return iter([])
+
+    G = _to_undirected_graph(G)
+
+    # FIXME: this zip() call appears to be the performance bottleneck for this
+    # function. Is there a better way?
+    (u, v) = zip(*ebunch)
+
+    try:
+        # Convert the ebunch lists to cupy arrays for passing to PLC, possibly
+        # mapping to integers if the Graph was renumbered.
+        # Allow the Graph renumber lookup (if renumbering was done) to check
+        # for invalid node IDs in ebunch.
+        u = G._list_to_nodearray(u)
+        v = G._list_to_nodearray(v)
+    except KeyError as n:
+        raise nx.NodeNotFound(f"Node {n} not in G.")
+    else:
+        # If G was not renumbered, then the ebunch nodes must be explicitly
+        # checked (note: ebunch can be very large). If not done,
+        # plc.jaccard_coefficients() will accept node IDs not in the graph and
+        # return a coefficient of 0 for them, which is not compatible with NX.
+        #
+        # FIXME: Is there a better way to do this?  Is there a utility to check
+        # if a node ID is valid for the graph?
+        if not hasattr(G, "key_to_id") or G.key_to_id is None:
+            ebunch_nodes = cp.unique(cp.concatenate([u, v]))
+            graph_nodes = cp.unique(
+                cp.concatenate(
+                    [
+                        G.src_indices,
+                        G.dst_indices,
+                        G._node_ids if G._node_ids is not None else cp.ndarray(0),
+                    ]
+                )
+            )
+            invalid = cp.setdiff1d(ebunch_nodes, graph_nodes, assume_unique=True)
+            if len(invalid) > 0:
+                raise nx.NodeNotFound(f"Node {invalid.tolist()[0]} not in G.")
```

Comment:
Thanks for being careful about behavior with missing nodes. There are no helper functions for this yet; each algorithm behaves differently and does it themselves. I think it probably makes sense to have a helper function.

Since we have the indices, we just need to know if any are out of bounds, so it's sufficient to check bounds with min/max.

Continuing from my `suggestion` code above (these aren't meant to be merged as-is), this could be:
```suggestion
    else:
        (u, v) = zip(*ebunch)
        try:
            src_indices = G._list_to_nodearray(u)
            dst_indices = G._list_to_nodearray(v)
        except KeyError as n:
            raise nx.NodeNotFound(f"Node {n} not in G.")
        if G.key_to_id is None and (
            (n := src_indices.max()) >= G._N
            or (n := dst_indices.max()) >= G._N
            or (n := src_indices.min()) < 0
            or (n := dst_indices.min()) < 0
        ):
            raise nx.NodeNotFound(f"Node {n} not in G.")  
```
Is `NodeNotFound` tested? If not, can we add a test to nx-cugraph or upstream?

Reply:
This LGTM.

> Is NodeNotFound tested? If not, can we add a test to nx-cugraph or upstream?

Yes, the NX tests have a few tests to ensure it's raised properly ([here](https://github.com/networkx/networkx/blob/f2c43bc6d22f470497ca2bd97c6458969ecf8320/networkx/algorithms/tests/test_link_prediction.py#L43)'s one) and they helped me discover a few edge cases IIRC.